File: sdk/elixir/lib/telecom_sdk/usage_api.ex
defmodule TelecomSDK.UsageAPI do
  @moduledoc """
  API for usage management
  """

  defstruct [:client]

  def new(client) do
    %__MODULE__{client: client}
  end

  def get_stats(api, subscriber_id, start_date, end_date) do
    params = %{
      start_date: DateTime.to_iso8601(start_date),
      end_date: DateTime.to_iso8601(end_date)
    }
    TelecomSDK.HTTPClient.get(api.client, "/v1/subscribers/#{subscriber_id}/usage", params)
  end

  def list_events(api, opts \\ []) do
    params = %{
      page: Keyword.get(opts, :page, 1),
      page_size: Keyword.get(opts, :page_size, 50)
    }

    params = if Keyword.get(opts, :subscriber_id), do: Map.put(params, :subscriber_id, Keyword.get(opts, :subscriber_id)), else: params
    params = if Keyword.get(opts, :usage_type), do: Map.put(params, :usage_type, Keyword.get(opts, :usage_type)), else: params
    params = if Keyword.get(opts, :start_date), do: Map.put(params, :start_date, DateTime.to_iso8601(Keyword.get(opts, :start_date))), else: params
    params = if Keyword.get(opts, :end_date), do: Map.put(params, :end_date, DateTime.to_iso8601(Keyword.get(opts, :end_date))), else: params

    TelecomSDK.HTTPClient.get(api.client, "/v1/usage/events", params)
  end

  def get_realtime(api, subscriber_id) do
    TelecomSDK.HTTPClient.get(api.client, "/v1/subscribers/#{subscriber_id}/realtime")
  end
end

File: sdk/ruby/lib/telecom_sdk/usage_api.rb
module TelecomSDK
  # API for usage management
  class UsageAPI
    def initialize(client)
      @client = client
    end

    def get_stats(subscriber_id, start_date, end_date)
      params = {
        start_date: start_date.iso8601,
        end_date: end_date.iso8601
      }
      @client.get("/v1/subscribers/#{subscriber_id}/usage", params)
    end

    def list_events(subscriber_id: nil, usage_type: nil, start_date: nil, end_date: nil, page: 1, page_size: 50)
      params = { page: page, page_size: page_size }
      params[:subscriber_id] = subscriber_id if subscriber_id
      params[:usage_type] = usage_type if usage_type
      params[:start_date] = start_date.iso8601 if start_date
      params[:end_date] = end_date.iso8601 if end_date
      @client.get("/v1/usage/events", params)
    end

    def get_realtime(subscriber_id)
      @client.get("/v1/subscribers/#{subscriber_id}/realtime")
    end
  end
end
